src/journal/JournalMetadata.cc

```diff
@@ -0,0 +1,464 @@
+}
+
+int JournalMetadata::unregister_client() {
+  assert(!m_client_id.empty());
+
+  ldout(m_cct, 10) << __func__ << ": " << m_client_id << dendl;
+  int r = client::client_unregister(m_ioctx, m_oid, m_client_id);
+  if (r < 0) {
+    lderr(m_cct) << "failed to unregister journal client '" << m_client_id
+                 << "': " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  notify_update();
+  return 0;
+}
+
+void JournalMetadata::add_listener(Listener *listener) {
+  Mutex::Locker locker(m_lock);
+  while (m_update_notifications > 0) {
+    m_update_cond.Wait(m_lock);
+  }
+  m_listeners.push_back(listener);
+}
+
+void JournalMetadata::remove_listener(Listener *listener) {
+  Mutex::Locker locker(m_lock);
+  while (m_update_notifications > 0) {
+    m_update_cond.Wait(m_lock);
+  }
+  m_listeners.remove(listener);
+}
+
+void JournalMetadata::set_minimum_set(uint64_t object_set) {
+  Mutex::Locker locker(m_lock);
+
+  ldout(m_cct, 20) << __func__ << ": current=" << m_minimum_set
+                   << ", new=" << object_set << dendl;
+  if (m_minimum_set >= object_set) {
+    return;
+  }
+
+  librados::ObjectWriteOperation op;
+  client::set_minimum_set(&op, object_set);
+
+  C_NotifyUpdate *ctx = new C_NotifyUpdate(this);
+  librados::AioCompletion *comp =
+    librados::Rados::aio_create_completion(ctx, NULL,
+                                           utils::rados_ctx_callback);
+  int r = m_ioctx.aio_operate(m_oid, comp, &op);
+  assert(r == 0);
+  comp->release();
+
+  m_minimum_set = object_set;
+}
+
+void JournalMetadata::set_active_set(uint64_t object_set) {
+  Mutex::Locker locker(m_lock);
+
+  ldout(m_cct, 20) << __func__ << ": current=" << m_active_set
+                   << ", new=" << object_set << dendl;
+  if (m_active_set >= object_set) {
+    return;
+  }
+
+  librados::ObjectWriteOperation op;
+  client::set_active_set(&op, object_set);
+
+  C_NotifyUpdate *ctx = new C_NotifyUpdate(this);
+  librados::AioCompletion *comp =
+    librados::Rados::aio_create_completion(ctx, NULL,
+                                           utils::rados_ctx_callback);
+  int r = m_ioctx.aio_operate(m_oid, comp, &op);
+  assert(r == 0);
+  comp->release();
+
+  m_active_set = object_set;
+}
+
+void JournalMetadata::flush_commit_position() {
+  {
+    Mutex::Locker locker(m_lock);
+    if (m_commit_position_task_ctx == NULL) {
+      return;
+    }
+
+    Mutex::Locker timer_locker(m_timer_lock);
+    m_timer->cancel_event(m_commit_position_task_ctx);
+    m_commit_position_task_ctx = NULL;
+  }
+  handle_commit_position_task();
+}
+
+void JournalMetadata::set_commit_position(
+    const ObjectSetPosition &commit_position, Context *on_safe) {
+  assert(on_safe != NULL);
+
+  Mutex::Locker locker(m_lock);
+  ldout(m_cct, 20) << __func__ << ": current=" << m_client.commit_position
+                   << ", new=" << commit_position << dendl;
+  if (commit_position <= m_client.commit_position ||
+      commit_position <= m_commit_position) {
```

Comment:
Good catch: the check against object_number shouldn't exist anymore.  It should only compare the tid ids since the tids are monotonically increasing.